### PR TITLE
Correlations: Add CorrelationSettings Page

### LINKED ISSUE #52027: Glue Correlations: Correlations settings page
As part of [#51323](https://github.com/grafana/grafana/issues/51323) we want to introduce a settings page where users can define correlations between their configured datasources. 

The page should sit in the configuration section (next to the Data Source settings page) and allow users to perform basic CRUD operations on correlations.

Data should be presented in an expandable table. When expanded, the table row should allow for modifications over the selected correlation.

We have multiple table-like components in the codebase, it would be nice to either reuse one of them or - if not possible - create a reusable table component (different form the Table viz component in `@grafana/ui`). GEX has one we could build on top of

### PR DESCRIPTION
**Pull request by Elfo404**
Adds a configuration page for correlations definitions
Fixes #52027

## Empty state

Screenshot 2022-08-18 at 10 31 43

## List of correlations
the list of defined correlations is presented in a table:
Screenshot 2022-08-18 at 10 32 42

## Editing correlations

https://user-images.githubusercontent.com/1170767/185362896-38bd46be-5081-4f70-88ea-19413af66f92.mov

> **Note**: read only correlations (where source datasource is read only) are still expandable, but the form is read only and there's no save button

Screenshot 2022-08-18 at 10 37 48

## Deleting correlations

> **Note**: delete button is not shown for read only correlations (where source datasource is read only)

https://user-images.githubusercontent.com/1170767/185363099-fdb92db0-0d7c-4609-acab-7eb948d4d4b2.mov

https://user-images.githubusercontent.com/1170767/185364166-ad5ee751-703f-4eb1-a896-83b6307b5d82.mov

## Notes:
https://github.com/grafana/grafana/pull/53821/commits/81d937af57b681eb1e4163c9a6f6fb40b4ccb4e8 adds the `readOnly` props to data returned by frontend settings

cc. @josiahg @radiorental 

## TODO:
- [x] add a proper icon. in the screenshots i'm using the actual icon but the PR does not include it because it looks like the process is about to change. waiting for https://github.com/grafana/grafana/pull/53766 to be merged
- [x] some types in CorrelationsPage and useCorrelations should be improved

### CODE REVIEW

**Comment by ifrost** [comment]
After adding a new correlation `GET correlations/` seems to be called twice:

Screen Shot 2022-08-19 at 15 17 30

File: public/app/features/correlations/CorrelationsPage.test.tsx [hunk 1]
```diff
@@ -0,0 +1,423 @@
...
+import { GrafanaContext } from 'app/core/context/GrafanaContext';
+import { contextSrv } from 'app/core/services/context_srv';
+import { configureStore } from 'app/store/configureStore';
+
+import { mockDataSource, MockDataSourceSrv } from '../alerting/unified/mocks';
+
+import CorrelationsPage from './CorrelationsPage';
+import { Correlation, CreateCorrelationParams } from './types';
+
+type DeepPartial<T> = T extends Array<infer U>
```
Comment (ifrost) [inline public/app/features/correlations/CorrelationsPage.test.tsx:19]: Isn't it `import {DeepPartial} from "react-hook-form";`? If not - could it be defined in some core utilities in grafana-data?

File: public/app/features/correlations/components/EmptyCorrelationsCTA.tsx [hunk 2]
```diff
@@ -0,0 +1,20 @@
+import React from 'react';
+
+import EmptyListCTA from 'app/core/components/EmptyListCTA/EmptyListCTA';
```
Comment (ifrost) [inline public/app/features/correlations/components/EmptyCorrelationsCTA.tsx:3]: Why with an expicit CTA suffix? Haven't seen it for any other call-to-action components.

File: public/app/core/utils/types.ts [hunk 3]
```diff
@@ -0,0 +1,3 @@
+type Truthy<T> = T extends false | '' | 0 | null | undefined ? never : T;
```
Comment (ifrost) [inline public/app/core/utils/types.ts:1]: Should we add it to grafana-data/types?

File: packages/grafana-ui/src/components/ConfirmButton/DeleteButton.tsx [hunk 4]
```diff
@@ -13,16 +13,19 @@ export interface Props {
   /** Disable button click action */
   disabled?: boolean;
   'aria-label'?: string;
+  /** Close after delete button is clicked */
```
Comment (ifrost) [inline packages/grafana-ui/src/components/ConfirmButton/DeleteButton.tsx:16]: Nit: redundant docs - it's just proxy to already documented property (in ConfirmButton.tsx)

File: public/app/features/correlations/CorrelationsPage.tsx [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+export default function CorrelationsPage() {
+  const navModel = useNavModel('correlations');
+  const [isAdding, setIsAdding] = useState(false);
+  const { correlations, create, remove, update, isLoading, error, reload } = useCorrelations();
+  const canWriteCorrelations = contextSrv.hasPermission(AccessControlAction.DataSourcesWrite);
+
+  const handleAdd = useCallback<ComponentProps<typeof AddCorrelationForm>['onSubmit']>(
+    async (correlation) => {
+      await create(correlation);
+      reload();
```
Comment (ifrost) [inline public/app/features/correlations/CorrelationsPage.tsx]: `create` already calls `reload`, isn't?

File: public/app/features/correlations/useCorrelations.ts [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  };
+
+  const remove = ({ sourceUID, uid }: RemoveCorrelationParams) => {
+    return backend.delete(`/api/datasources/uid/${sourceUID}/correlations/${uid}`).then(reload);
+  };
+
+  const update = ({ sourceUID, uid, ...correlation }: UpdateCorrelationParams) => {
+    return backend.patch(`/api/datasources/uid/${sourceUID}/correlations/${uid}`, correlation).then((data) => {
+      reload();
+      return toEnrichedCorrelationData(data);
```
Comment (ifrost) [inline public/app/features/correlations/useCorrelations.ts:70]: Does it actually need to return anything? Doesn't seem to be used, wouldn't in this case `reload` revalidate the data to most recent?

File: public/app/features/correlations/CorrelationsPage.tsx (hunk 5 above)
Comment (ifrost) [reply public/app/features/correlations/CorrelationsPage.tsx]: Sorry, commented after you changed it. NVM.

Comment (Elfo404) [reply public/app/features/correlations/CorrelationsPage.tsx]: this is why https://github.com/grafana/grafana/pull/53821#issuecomment-1220671286 was happening, removed in https://github.com/grafana/grafana/pull/53821/commits/fbf708d3f287fc29323a94d1b4e3c82dda9d0d94

File: public/app/core/utils/types.ts (hunk 3 above)
Comment (Elfo404) [reply public/app/core/utils/types.ts:1]: it is not exported by grafana-data, i just moved it from the ES datasource to somewhere more sensible to be reused, i don't think this and `isTruthy` belongs to @grafana/data

**Comment by ifrost** [comment]
Out of curiosity - why `GET /correlations` returns 404 when there are no correlations defined instead of an empty list? Not sure if we use such a pattern in other endpoints.

**Comment by ifrost** [comment]
We definitely need a better validation for provisioned data sources but it can be added in one of the upcoming PRs. We should make it clear that provisioned data source cannot be used as the source (the provisioning file should be used instead).

Screen Shot 2022-08-22 at 10 55 35

File: packages/grafana-ui/src/components/ConfirmButton/DeleteButton.tsx (hunk 4 above)
Comment (Elfo404) [reply packages/grafana-ui/src/components/ConfirmButton/DeleteButton.tsx:16]: yes, but you still need to add it here to get the description in - for example - vscode

File: public/app/features/correlations/CorrelationsPage.test.tsx (hunk 1 above)
Comment (Elfo404) [reply public/app/features/correlations/CorrelationsPage.test.tsx:19]: 🙈 forgot to remove that one, thanks!

File: public/app/features/correlations/components/EmptyCorrelationsCTA.tsx (hunk 2 above)
Comment (Elfo404) [reply public/app/features/correlations/components/EmptyCorrelationsCTA.tsx:3]: no real reason, i just used the component's name when importing given it's a default export.

as of why i'm exporting it as `EmptyCorrelationsCTA` it's kinda for the same reason.

File: public/app/features/correlations/CorrelationsPage.test.tsx (hunk 1 above)
Comment (Elfo404) [reply public/app/features/correlations/CorrelationsPage.test.tsx]: 163bb188c08f06e54f752cac8d174c2b8b12873e

**Comment by Elfo404** [comment]
> Out of curiosity - why `GET /correlations` returns 404 when there are no correlations defined instead of an empty list? Not sure if we use such a pattern in other endpoints.

@ifrost i clearly overlooked that, i think it should return 200 with an empty list (note the empty list  of correlations is still returned)

**Comment by Elfo404** [comment]
> We definitely need a better validation for provisioned data sources but it can be added in one of the upcoming PRs. We should make it clear that provisioned data source cannot be used as the source (the provisioning file should be used instead).
> 
> Screen Shot 2022-08-22 at 10 55 35

I agree, i think it can also be included in this PR TBH, would something like this be better in your opinion?
Screenshot 2022-08-22 at 11 23 33

File: public/app/features/correlations/useCorrelations.ts (hunk 6 above)
Comment (Elfo404) [reply public/app/features/correlations/useCorrelations.ts:70]: it's true that right now we are not using that returned data, but the hook itself shouldn't make any assumption on how their consumers use it IMO. in this case we do not use the response from this call (or all the others aside `get`) but the API returns that data, which may be useful to consumers, especially given we'll have to expose this and use it in other parts of the codebase

**Comment by ifrost** [comment]
> > We definitely need a better validation for provisioned data sources but it can be added in one of the upcoming PRs. We should make it clear that provisioned data source cannot be used as the source (the provisioning file should be used instead).
> > Screen Shot 2022-08-22 at 10 55 35
> 
> I agree, i think it can also be included in this PR TBH, would something like this be better in your opinion? Screenshot 2022-08-22 at 11 23 33

Looks great! If it's easy to add go for it, but can be added later as well.

**Comment by Elfo404** [comment]
> Looks great! If it's easy to add go for it, but can be added later as well.

47b96b8477 :)

File: public/app/features/correlations/CorrelationsPage.tsx [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+  return (
+    <Page navModel={navModel}>
+      <Page.Contents>
+        <div>
+          <HorizontalGroup justify="space-between">
+            <div>
+              <h4>Correlations</h4>
+              <p>Define how data living in different data sources relates to each other.</p>
+            </div>
+            {canWriteCorrelations && data?.length !== 0 && data !== undefined && !isAdding && (
```
Comment (ifrost) [inline public/app/features/correlations/CorrelationsPage.tsx:102]: It doesn't seem to work - the button is still visible after clicking on it:


https://user-images.githubusercontent.com/745532/186107837-954e19dd-4111-454b-9edc-c5993b69f62e.mov

File: public/app/features/correlations/components/Table/index.tsx [hunk 8]
```diff
@@ -0,0 +1,173 @@
...
+    }
+  `,
+  th: css`
+    position: relative;
+  `,
+  sortIcon: css`
+    position: absolute;
+    width: 16px;
+    overflow: hidden;
+  `,
```
Comment (ifrost) [inline public/app/features/correlations/components/Table/index.tsx:39]: What's the use case for relative/absolute positioning?

File: public/app/features/correlations/CorrelationsPage.tsx (hunk 7 above)
Comment (Elfo404) [reply public/app/features/correlations/CorrelationsPage.tsx:102]: this does a different thing, it removes the "add new", not the submit button

File: public/app/features/correlations/components/Table/index.tsx (hunk 8 above)
Comment (Elfo404) [reply public/app/features/correlations/components/Table/index.tsx:39]: probably a leftover, will remove 👍

File: public/app/features/correlations/CorrelationsPage.tsx (hunk 7 above)
Comment (Elfo404) [reply public/app/features/correlations/CorrelationsPage.tsx:102]: (working on it though, hold on :D )

File: public/app/features/correlations/Forms/AddCorrelationForm.tsx [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+    padding: ${theme.spacing(1)};
+    margin-bottom: ${theme.spacing(2)};
+  `,
+  linksToContainer: css`
+    flex-grow: 1;
+    /* This is the width of the textarea minus the sum of the label&description fields,
+     * so that this element takes exactly the remaining space and the inputs will be
+     * nicely aligned with the textarea
+    **/
+    max-width: ${theme.spacing(80 - 64)};
```
Comment (ifrost) [inline public/app/features/correlations/Forms/AddCorrelationForm.tsx:29]: We may need something more robust or different layouting as it may not work well with i18n but we can revise it later.

**Review by ifrost — APPROVED** [review]
🎉

File: public/app/features/correlations/CorrelationsPage.tsx [hunk 10]
```diff
@@ -0,0 +1,215 @@
...
+        {!data && get.loading && (
+          <div className={loaderWrapper}>
+            <LoadingPlaceholder text="loading..." />
+          </div>
+        )}
+
+        {showEmptyListCTA && <EmptyCorrelationsCTA onClick={() => setIsAdding(true)} />}
+
+        {
+          // This error is not actionable, it'd be nice to have a recovery button
```
Comment (L-M-K-B) [inline public/app/features/correlations/CorrelationsPage.tsx:138]: Is this worthy to be solved right here or in another PR? Would you like to create an issue for it?

File: public/app/features/correlations/CorrelationsPage.tsx [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+    );
+  },
+  ({ cell: { value } }, { cell: { value: prevValue } }) => {
+    return value.type === prevValue.type && value.name === prevValue.name;
+  }
+);
+
+const noWrap = css`
+  white-space: nowrap;
+`;
```
Comment (L-M-K-B) [inline public/app/features/correlations/CorrelationsPage.tsx:202]: In order to have all styling in one place I'd suggest to add this below line 179.

File: public/app/features/correlations/components/Table/index.tsx [hunk 12]
```diff
@@ -0,0 +1,161 @@
...
+  data: TableData[];
+  renderExpandedRow?: (row: TableData) => JSX.Element;
+  className?: string;
+  getRowId: TableOptions<TableData>['getRowId'];
+}
+
+/**
+ * non-viz table component.
+ * Will need most likely to be moved in @grafana/ui
+ */
```
Comment (L-M-K-B) [inline public/app/features/correlations/components/Table/index.tsx:65]: Why aren't you adding it there in this PR? In case this isn't possible right now are you going to open an issue for it?

**Review by L-M-K-B — COMMENTED** [review]
I have a more general question regarding the TODOs in this PR: Do you intent to open issues for them in order to follow up on them later? If not, how do you make sure that they are getting solved?

**Comment by ifrost** [comment]
> I have a more general question regarding the TODOs in this PR: Do you intent to open issues for them in order to follow up on them later? If not, how do you make sure that they are getting solved?

In my opinion, a TODO in the code should always have a linked issue for example like here https://github.com/grafana/grafana/blob/main/public/app/plugins/datasource/loki/components/LokiLabelBrowser.tsx#L31. Otherwise it never gets done or the context of the todo is lost in a few weeks.

File: public/app/features/correlations/CorrelationsPage.tsx (hunk 11 above)
Comment (Elfo404) [reply public/app/features/correlations/CorrelationsPage.tsx:202]: i do not agree. there are 3 styles objects/functions in this file, each one is defined just before the component using them.

File: public/app/features/correlations/components/Table/index.tsx (hunk 12 above)
Comment (Elfo404) [reply public/app/features/correlations/components/Table/index.tsx:65]: because this is still in WIP, we are working with the UX  team and on a design doc right now to define what needs to be done for it to be in the ui package. there's already a draft issue in our board for this.

File: public/app/features/correlations/Forms/AddCorrelationForm.tsx (hunk 9 above)
Comment (Elfo404) [reply public/app/features/correlations/Forms/AddCorrelationForm.tsx:29]: yeah, but tbh the whole form layout is temporary as we don't have anything else there. the whole thing will change once we add all the correlations settings in the form.

File: public/app/features/correlations/CorrelationsPage.tsx (hunk 10 above)
Comment (Elfo404) [reply public/app/features/correlations/CorrelationsPage.tsx:138]: https://github.com/grafana/grafana/issues/54288
